### PR TITLE
8334258: Compiler erronousely allows access to instance variable in argument expression of a constructor invocation

### PR DESCRIPTION
**Pull request by archiecobbs**
For the "flexible constructors" JEP 482 feature, the compiler allows assignments to instance fields in early construction contexts. However, when it creates this exception it is failing to check that the feature is enabled. This patch adds that check.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8334258](https://bugs.openjdk.org/browse/JDK-8334258): Compiler erronousely allows access to instance variable in argument expression of a constructor invocation (**Bug** - P3)

### Reviewers
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**) ⚠️ Review applies to [7abe6171](https://git.openjdk.org/jdk/pull/19712/files/7abe6171e4e62df660185de8856f200498385a12)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19712/head:pull/19712` \
`$ git checkout pull/19712`

Update a local copy of the PR: \
`$ git checkout pull/19712` \
`$ git pull https://git.openjdk.org/jdk.git pull/19712/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19712`

View PR using the GUI difftool: \
`$ git pr show -t 19712`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19712.diff">https://git.openjdk.org/jdk/pull/19712.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19712#issuecomment-2166794695)

### CODE REVIEW

File: test/langtools/tools/javac/SuperInit/EarlyAssignmentNoPreview3.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ * @bug 8334258
+ * @summary Disallow early assignment if FLEXIBLE_CONSTRUCTORS preview feature is not enabled
+ * @compile/fail/ref=EarlyAssignmentNoPreview3.out -XDrawDiagnostics EarlyAssignmentNoPreview3.java
+ */
+public class EarlyAssignmentNoPreview3 {
+
+    Runnable r;
+
+    public EarlyAssignmentNoPreview3() {
+        this(EarlyAssignmentNoPreview3.this.r = () -> System.out.println("hello"));
```
Comment (mcimadamore) [inline test/langtools/tools/javac/SuperInit/EarlyAssignmentNoPreview3.java:12]: Optional suggestion: you could in principle consolidate these tests by overloading the constructor - e.g. you can have
```Foo(int)
Foo(int, int)
Foo(int, int, int)```
And each of those could delegate to `Foo(Runnable)`, with slightly different expressions

**Review by mcimadamore — APPROVED** [review]

File: test/langtools/tools/javac/SuperInit/EarlyAssignmentNoPreview3.java (hunk 1 above)
Comment (archiecobbs) [reply test/langtools/tools/javac/SuperInit/EarlyAssignmentNoPreview3.java:12]: I tried that, but you only get one "feature disabled" error per file so it doesn't work to put multiple violations in the same file.

**Comment by mcimadamore** [comment]
@archiecobbs  There seem to be related test failures in the github actions - please double check

**Comment by mcimadamore** [comment]
@archiecobbs - you might want to delete or edit your "integrate" comment, so that this PR won't be accidentally integrated/sponsored.

**Comment by archiecobbs** [comment]
> you might want to delete or edit your "integrate" comment, so that this PR won't be accidentally integrated/sponsored.

Thanks - done. Learned a new skara command in the process.

**Comment by mcimadamore** [comment]
Question: is this ready to go? E.g. this should not have any dependencies, correct?

**Comment by archiecobbs** [comment]
> Question: is this ready to go? E.g. this should not have any dependencies, correct?

Yes, this fix should be independent of the other changes floating around. Thanks.

**Pull request closed without merging**